aequilibrae/project/network/link.py

```diff
@@ -1,25 +1,194 @@
+                if key not in self.__original__:
+                    continue
+                if val != self.__original__[key]:
+                    if key == 'geometry' and val is not None:
+                        data.append(val.wkb)
+                    else:
+                        data.append(val)
+                    txts.append(f'"{key}"=?')
+
+            if not data:
+                logger.warn(f'Nothing to update for link {self.link_id}')
+                return
+
+            txts = ','.join(txts) + ' where link_id=?'
+            data.append(self.link_id)
+            sql = f'Update Links set {txts}'
+
+        logger.error(sql)
+        curr.execute(sql, data)
+        conn.commit()
+        conn.close()
+        self.__new = False
+
+    def set_modes(self, modes: str):
+        """Sets the modes acceptable for this link
+
+        Args:
+            *modes* (:obj:`str`): string with all mode_ids to be assigned to this link
+        """
+
+        if not isinstance(modes, str):
+            raise ValueError('Modes field needs to be a string')
+        if modes == '':
+            raise ValueError('Modes field needs to have at least one mode')
+
+        self.__dict__["modes"] = modes
+
+    def add_mode(self, mode: [str, Mode]):
+        """Adds a new mode to this link
+
+        Raises a warning if mode is already allowed on the link, and fails if mode does not exist
+
+        Args:
+            *mode_id* (:obj:`str` or `Mode`): Mode_id of the mode or mode object to be added to the link
+        """
+        mode_id = self.__validate(mode)
+
+        if mode_id in self.modes:
+            logger.warn('Mode already active for this link')
+            return
+
+        self.__dict__["modes"] += mode_id
+
+    def drop_mode(self, mode: [str, Mode]):
+        """Removes a mode from this link
+
+        Raises a warning if mode is already NOT allowed on the link, and fails if mode does not exist
+
+        Args:
+            *mode_id* (:obj:`str` or `Mode`): Mode_id of the mode or mode object to be removed from the link
+        """
+
+        mode_id = self.__validate(mode)
+
+        if mode_id not in self.modes:
+            logger.warn('Mode already inactive for this link')
+            return
+
+        if len(self.modes) == 1:
+            raise ValueError('Link needs to have at least one mode')
+
+        self.__dict__['modes'] = self.modes.replace(mode_id, '')
+
+    def data_fields(self) -> list:
+        """lists all data fields for the link, as available in the database
+
+        Returns:
+            *data fields* (:obj:`list`): list of all fields available for editing
+        """
+
+        return list(self.__original__.keys())
+
+    def __validate(self, mode: [str, Mode]) -> str:
+        if isinstance(mode, Mode):
+            mode_id = mode.mode_id
+        elif isinstance(mode, str):
+            if len(mode) > 1:
+                raise ValueError('A mode_id is a single character')
```

Comment:
have you thought about having a validation stage that collects errors for all the links, reports them and then raises an Error?

Reply:
What do you mean?  This is only being raised if the user tries to alter the characteristics of individual links.

Reply:
I meant separating the manipulation from the saving... I guess I'm thinking of users making a lot of edits in a script and then being able to get an overview of which parts of that didn't validate, rather than 

1. Run the edit script,
2. Get an exception, fix that link edit
3. Re-run
4. Get a diff exception
5. Fix that.
6. Go to 3

Reply:
I see.  I had not considered that issue.  I can do that in the Links class with an overarching save() class

Reply:
might not be a real issue though...

Reply:
`def save(allow_fail=True):
  errors = []
  for link in self.__items.values()
    try:
        link.save()
    except Exception as e:
         errors.append([link.link_id, e.args])
    
  for link_id, err in errors:
    logger.error(f'link {link_id} was not saved due to error: {err}'

  if errors and allow_fail:
     raise Exception('saving links returned an error')

`

Reply:
I know it's pythonic, but raise exceptions just to catch them goes against every bone in my body 

Arguments against exceptions: 
* https://softwareengineering.stackexchange.com/questions/189222/are-exceptions-as-control-flow-considered-a-serious-antipattern-if-so-why
* https://web.archive.org/web/20140430044213/http://c2.com/cgi-bin/wiki?DontUseExceptionsForFlowControl

Why Python is a special snowflake:
* https://stackoverflow.com/questions/16138232/is-it-a-good-practice-to-use-try-except-else-in-python

Reply:
Some errors will be raised in the database, so the only way to collect all errors and fail only once is to try-except

Reply:
I'll leave the idea of collecting all errors aside for now, then